File: Web/tests/support/recordingAudioEngine.ts
import type { AudioEngine } from '../../src/domain/protocols/audioEngine.js'
import type { Tone } from '../../src/domain/instrument/tone.js'
import type { ToneChange } from '../../src/domain/playing/toneChange.js'
import type { ReedRelease } from '../../src/domain/playing/reedRelease.js'
import type { BreathIntensity } from '../../src/domain/playing/breathIntensity.js'
import type { BendDepth } from '../../src/domain/playing/bendDepth.js'
import type { VibratoDepth } from '../../src/domain/playing/vibratoDepth.js'
import type { CupDepth } from '../../src/domain/playing/cupDepth.js'

export class RecordingAudioEngine implements AudioEngine {
    readonly soundedTones: (readonly Tone[])[] = []
    readonly toneChanges: ToneChange[] = []
    readonly intensities: BreathIntensity[] = []
    readonly bends: BendDepth[] = []
    readonly vibratos: VibratoDepth[] = []
    readonly cups: CupDepth[] = []
    readonly releases: ReedRelease[] = []

    get silencings(): number {
        return this.releases.length
    }

    get hertzOfTheFirstTone(): number {
        return this.soundedTones[0]?.[0]?.hertz ?? 0
    }

    get hertzOfTheLastTone(): number {
        return this.soundedTones[this.soundedTones.length - 1]?.[0]?.hertz ?? 0
    }

    get hertzOfEveryTone(): number[] {
        return this.soundedTones.map(tones => Math.round(tones[0]?.hertz ?? 0))
    }

    async prepare(): Promise<void> {}

    soundTones(tones: readonly Tone[], change: ToneChange): void {
        this.soundedTones.push(tones)
        this.toneChanges.push(change)
    }

    changeIntensity(intensity: BreathIntensity): void {
        this.intensities.push(intensity)
    }

    changeBend(depth: BendDepth): void {
        this.bends.push(depth)
    }

    changeVibrato(depth: VibratoDepth): void {
        this.vibratos.push(depth)
    }

    cupHands(depth: CupDepth): void {
        this.cups.push(depth)
    }

    silence(release: ReedRelease): void {
        this.releases.push(release)
    }
}

File: Core/Sources/HarmonicaCoreTestSupport/RecordingAudioEngine.swift
import HarmonicaCore

@MainActor
public final class RecordingAudioEngine: AudioEngineProtocol {
    public private(set) var soundedTones: [[Tone]] = []
    public private(set) var toneChanges: [ToneChange] = []
    public private(set) var intensities: [BreathIntensity] = []
    public private(set) var bends: [BendDepth] = []
    public private(set) var vibratos: [VibratoDepth] = []
    public private(set) var cups: [CupDepth] = []
    public private(set) var releases: [ReedRelease] = []
    public var preparationFailure: AudioEngineError?

    nonisolated public init() {}

    public var hertzOfTheFirstTone: Double {
        hertz(of: soundedTones.first)
    }

    public var hertzOfTheLastTone: Double {
        hertz(of: soundedTones.last)
    }

    public func prepare() async throws(AudioEngineError) {
        guard let preparationFailure else { return }

        throw preparationFailure
    }

    public func soundTones(_ tones: [Tone], as change: ToneChange) {
        soundedTones.append(tones)
        toneChanges.append(change)
    }

    public func changeIntensity(to intensity: BreathIntensity) {
        intensities.append(intensity)
    }

    public func changeBend(to depth: BendDepth) {
        bends.append(depth)
    }

    public func changeVibrato(to depth: VibratoDepth) {
        vibratos.append(depth)
    }

    public func cupHands(to depth: CupDepth) {
        cups.append(depth)
    }

    public func silence(_ release: ReedRelease) {
        releases.append(release)
    }

    private func hertz(of tones: [Tone]?) -> Double {
        tones?.first?.pitch.converted(to: .hertz).value ?? 0
    }
}
